jgrapht-core/src/main/java/org/jgrapht/alg/vertexcover/TreeVCImpl.java

```diff
@@ -0,0 +1,193 @@
+     * </p>
+     *
+     * <p>
+     *      Note: The vertex cover is computed lazily, when needed.
+     * </p>
+     *
+     * @param graph the input graph
+     * @param root the input root
+     * @throws NullPointerException if {@code graph} is {@code null}
+     * @throws NullPointerException if {@code root} is {@code null}
+     * @throws IllegalArgumentException if {@code root} is an invalid vertex
+     */
+    public TreeVCImpl(Graph<V, E> graph, V root){
+        this(graph, Collections.singleton(Objects.requireNonNull(root, "root cannot be null")));
+    }
+
+    /**
+     * Creates a new TreeVCImpl instance.
+     *
+     * <p>
+     *      Note: The constructor does NOT check if the input graph is a valid tree.
+     * </p>
+     *
+     * <p>
+     *     Note: If two roots correspond to the same tree, an error will be thrown.
+     * </p>
+     *
+     * <p>
+     *      Note: The vertex cover is computed lazily, when needed.
+     * </p>
+     *
+     * @param graph the input graph
+     * @param roots the input root
+     * @throws NullPointerException if {@code graph} is {@code null}
+     * @throws NullPointerException if {@code roots} is {@code null}
+     * @throws IllegalArgumentException if {@code roots} is empty
+     * @throws IllegalArgumentException if {@code roots} contains an invalid vertex
+     */
+    public TreeVCImpl(Graph<V, E> graph, Set<V> roots) {
+        this.graph = Objects.requireNonNull(graph, "graph cannot be null");
+        this.roots = Objects.requireNonNull(roots, "set of roots cannot be null");
+
+        if (this.roots.isEmpty())
+            throw new IllegalArgumentException("set of roots cannot be empty");
+
+        if (!this.graph.vertexSet().containsAll(roots))
+            throw new IllegalArgumentException("invalid set of roots");
+    }
+
+    // Populate parent map by iterating through the BFS tree of root
+    private void bfs(V root, Set<V> visited, Map<V, V> parent){
+        BreadthFirstIterator<V, E> bfs = new BreadthFirstIterator<>(graph, root);
+
+        while (bfs.hasNext()) {
+            V u = bfs.next();
+            visited.add(u);
+            V p = bfs.getParent(u);
+
+            if (p != null)
+                parent.put(u, p);
+        }
+    }
+
+    private void computeMinimumVertexCover(){
+        Map<V, V> parent =  new HashMap<>();
+        Set<V> visited = new HashSet<>();
+
+        for (V root: roots)
+            if (!visited.contains(root))
+                bfs(root, visited, parent);
+            else
+                throw new IllegalArgumentException("multiple roots in the same tree");
+
+        Set<V> leaves = new HashSet<>();
+        Set<V> vc = new HashSet<>();
+        Set<V> deleted = new HashSet<>();
+
+        // A vertex v is a leaf iff there is no vertex u such that v is the parent of u
+        for (V v: graph.vertexSet())
+            if (!parent.values().contains(v))
+                leaves.add(v);
+        do {
```

Comment:
Instead of actually deleting nodes, wouldn't the following much shorted code be equivalent?

```
while(!leaves.isEmpty()){
	Set<V> parents=new HashSet<>();
	for(V v leaf:leaves){
		V parent=parentMap.get(leaf);
		if(parent !=null && !vc.contains(parent){
			vc.add(parent);
			parents.add(parent); //only add parent if we haven't reached this node via another branch already
		}
	}
	leaves=parents;
}
```